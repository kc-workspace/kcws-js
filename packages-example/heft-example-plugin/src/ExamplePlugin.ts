import type {
  HeftConfiguration,
  IHeftTaskPlugin,
  IHeftTaskSession,
} from "@rushstack/heft";

import { PLUGIN_NAME } from "./shared";

/**
 * The options for esbuild build() function.
 *
 * @public
 */
export interface IExampleOption {
  /** The example option field */
  example?: string;
}

class ExamplePlugin implements IHeftTaskPlugin<IExampleOption> {
  public static DEFAULT_MESSAGE: string = `Using example to do nothing`;
  private _accessor: undefined;

  public get accessor(): undefined {
    return this._accessor;
  }

  public apply(
    session: IHeftTaskSession,
    _heftConfiguration: HeftConfiguration,
    _options?: IExampleOption | undefined
  ): void {
    session.hooks.run.tapPromise(PLUGIN_NAME, async () => {
      session.logger.terminal.writeLine(ExamplePlugin.DEFAULT_MESSAGE);
      // Your logic should be here...
    });
  }
}

export default ExamplePlugin;
