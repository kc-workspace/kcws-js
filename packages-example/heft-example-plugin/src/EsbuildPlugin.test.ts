import type {
  HeftConfiguration,
  IHeftParameters,
  IHeftTaskHooks,
  IHeftTaskRunHookOptions,
  IHeftTaskRunIncrementalHookOptions,
  IHeftTaskSession,
} from "@rushstack/heft";

import { AsyncParallelHook } from "tapable";
import { StringBufferTerminalProvider, Terminal } from "@rushstack/terminal";
import { MockScopedLogger } from "@rushstack/heft/lib/pluginFramework/logging/MockScopedLogger";

import ExamplePlugin from "./ExamplePlugin";

const mockTaskSession = (
  parameters?: Partial<IHeftParameters>
): { session: IHeftTaskSession; terminal: StringBufferTerminalProvider } => {
  const terminalProvider = new StringBufferTerminalProvider(false);
  const terminal = new Terminal(terminalProvider);
  const logger = new MockScopedLogger(terminal);

  const _parameters: IHeftParameters = Object.assign(
    { watch: false },
    parameters
  ) as IHeftParameters;

  const session = {
    logger,
    parameters: _parameters,
    hooks: {
      run: new AsyncParallelHook<IHeftTaskRunHookOptions>(),
      runIncremental:
        new AsyncParallelHook<IHeftTaskRunIncrementalHookOptions>(),
    } as IHeftTaskHooks,
    parsedCommandLine: undefined!,
    requestAccessToPluginByName: undefined!,
    taskName: "esbuild",
    tempFolderPath: "/tmp/test",
  };
  return { session, terminal: terminalProvider };
};

const mockHeftConfiguration = (
  config?: Partial<HeftConfiguration>
): HeftConfiguration => {
  const base = {
    buildFolderPath: "/tmp/build",
    projectPackageJson: {
      name: "example",
      version: "v0.1.0",
    },
  } as HeftConfiguration;

  return Object.assign(base, config);
};

describe("Example heft-plugin", () => {
  it("should create with parameterless constructor", () => {
    const plugin = new ExamplePlugin();
    expect(plugin).not.toBeFalsy();
  });

  it("should return undefined as accessor", () => {
    const plugin = new ExamplePlugin();
    expect(plugin.accessor).toBeUndefined();
  });

  it("should do something", async () => {
    const plugin = new ExamplePlugin();
    const { session, terminal } = mockTaskSession();
    const config = mockHeftConfiguration();

    plugin.apply(session, config);
    await session.hooks.run.promise(undefined!);

    expect(terminal.getOutput()).toMatchSnapshot();
  });
});
