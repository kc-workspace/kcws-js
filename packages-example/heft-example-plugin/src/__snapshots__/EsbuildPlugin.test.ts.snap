// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example heft-plugin should do something 1`] = `"Using example to do nothing[n]"`;
